sonar/modules/documents/dnb.py

```diff
@@ -0,0 +1,118 @@
+
+
+class Dnb:
+    """Dnb class."""    
+
+    @classmethod
+    def headers(cls):
+        """."""
+        username = current_app.config.get('SONAR_APP_URN_DNB_USERNAME')
+        password = current_app.config.get('SONAR_APP_URN_DNB_PASSWORD')
+        string = f"{username}:{password}"
+        auth = base64.b64encode(string.encode()).decode('ascii')
+        return {
+            'Authorization': f'Basic {auth}',
+            'Content-Type': 'application/json',
+            'Accept': 'application/json'
+        }
+
+    @classmethod
+    def verify(cls, urn_code):
+        """Verify the existence for a URN.
+
+        :param urn_code: the urn code.
+        :returns: True if urn is registered, otherwise False.
+        """
+        # Documentation: https://wiki.dnb.de/display/URNSERVDOK/URN-Service+API
+        # https://wiki.dnb.de/display/URNSERVDOK/Beispiele%3A+URN-Verwaltung
+        answer = False
+        try:
+            response = requests.request(
+                'HEAD',
+                f"{current_app.config.get('SONAR_APP_URN_DNB_BASE_URL')}"\
+                f"/urns/urn/{urn_code}",
+                headers=cls.headers()
+            )
+            answer = response.status_code == 200
+        except Exception as error:
+            pass
+        return answer
+
+
+    @classmethod
+    def register_document(cls, document):
+        """Register a new URN code.
+
+        :param document: the document.
+        :returns: True if urn is registered, otherwise False.
+        """
+        from sonar.modules.documents.api import DocumentRecord
+        from sonar.modules.organisations.api import OrganisationRecord
+        answer = False
+        sonar_base_url = current_app.config.get('SONAR_APP_BASE_URL')
+        document = DocumentRecord(document)
+        if orgs := document.get('organisation', []):
+            # In case of multiple organisations we redirect to the global view
+            org_code = current_app.config.get('SONAR_APP_DEFAULT_ORGANISATION')        
+            if org := orgs[0]:
+                org = OrganisationRecord(org)
+                org = org.replace_refs()
+                if org.get('isDedicated') or org.get('isShared'):
+                    org_code = org.get('code')
+            url = f"{sonar_base_url}/{org_code}/documents/{document.get('pid')}"
+            data = {
+                'urn': document.get_rero_urn_code(document),
+                'urls': [
+                    {
+                        'url': url,
+                        'priority': 1
+                    }
+                ]
+            }
+            try:
+                response = requests.request(
+                    'POST',
+                    f"{current_app.config.get('SONAR_APP_URN_DNB_BASE_URL')}/urns",
+                    headers=cls.headers(),
+                    data=json.dumps(data)
+                )
+                answer = response.status_code == 201
+            except Exception as error:
```

Comment:
Should we log the error? Otherwise use: https://docs.sourcery.ai/refactorings/use-contextlib-suppress/